source/docs/guide/src/calc.md

```diff
@@ -0,0 +1,76 @@
+# Structured Proofs by Calculation
+
+## Motivation
+
+Sometimes, you need to establish some relation `R` between two expressions, say,
+`a_1` and `a_n`, where it might be easier to do this in a series of steps, `a_1`
+to `a_2`, `a_2` to `a_3`, ... all the way to `a_n`. One might do this by just
+doing all the steps at once, but as mentioned in the section on
+[assert-by](./assert_by.md), a better approach might be to split it into a
+collection of restricted contexts. This is better, but still might not be ideal,
+since you need to repeat each of the intermediate expressions at each point.
+
+## `calc!`ulations, to Reduce Redundant Redundancy
+
+The `calc!` macro supports structured proofs through calculations.
+
+In particular, one can show `a_1 R a_n` for some transitive relation `R` by performing a series
+of steps `a_1 R a_2`, `a_2 R a_3`, ... `a_{n-1} R a_n`. The calc macro provides both convenient
+syntax sugar to perform such a proof conveniently, without repeating oneself too often, or
+exposing the internal steps to the outside context.
+
+The expected usage looks like:
+
+```rust
+calc! {
+  (R)
+  a_1; { /* proof that a_1 R a_2 */ }
+  a_2; { /* proof that a_2 R a_3 */ }
+   ...
+  a_n;
+}
+```
+
+Notice that you mention of `a_1`, `a_2`, ... `a_n` only once each. Additionally,
+the proofs for each of the steps is localized, and restricted to only its
+particular step, ensuring that proof context is not polluted.
+
+The body of the function where this `calc` statement is written only gets to see
+`a_1 R a_n`, and not any of the intermediate steps (or their proofs), further
+limiting proof context pollution.
+
+Currently, the `calc!` macro supports common transitive relations for `R` (such
+as `==` and `<=`). This set of relations may be extended in the future.
```

Comment:
Just out of curiosity, does it support `=~=` and `=~~=`?

Reply:
At the moment, no; `calc` is older than `=~=` and `=~~=`.  Adding it is not difficult, but there is an interesting decision to be made as to what it compiles to: should saying R=`=~=` mean `=~=` _everywhere_ or should it mean `==` outside and all the internal ones being `=~=`? The former is less surprising wrt expansion itself (for anyone reading the `calc` code), but the latter would be more stable (and lead to less context pollution, and also the fact that we're restricting it should be practically undetectable by any user of the feature). I like the latter, but it requires a bit more work to implement than the former, so I'm holding off on that for a bit while I think through which of the designs work better.